openquake/hazardlib/gsim/afshari_stewart_2016.py

```diff
@@ -0,0 +1,226 @@
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU Affero General Public License for more details.
+#
+# You should have received a copy of the GNU Affero General Public License
+# along with OpenQuake. If not, see <http://www.gnu.org/licenses/>.
+
+"""
+Module exports :class:`AfshariStewart2016`,
+               :class:`AfshariStewart2016Japan`
+"""
+from __future__ import division
+
+import numpy as np
+
+from openquake.hazardlib.gsim.base import CoeffsTable, GMPE
+from openquake.hazardlib import const
+from openquake.hazardlib.imt import RSD595, RSD575, RSD2080
+
+
+class AfshariStewart2016(GMPE):
+    """
+    Implements the GMPE of Afshari & Stewart (2016) for relative significant
+    duration for 5 - 75 %, 5 - 95 % and 20 - 80 % Arias Intensity.
+
+    Afshari, K. and Stewart, J. P. (2016) "Physically Parameterized Prediction
+    Equations for Signficant Duration in Active Crustal Regions", Earthquake
+    Spectra, 32(4), 2057 - 2081
+    """
+    #: Supported tectonic region type is active shallow crust
+    DEFINED_FOR_TECTONIC_REGION_TYPE = const.TRT.ACTIVE_SHALLOW_CRUST
+
+    #: Supported intensity measure types are 5 - 95 % Arias and 5 - 75 % Arias
+    #: significant duration
+    DEFINED_FOR_INTENSITY_MEASURE_TYPES = set([
+        RSD595,
+        RSD575,
+        RSD2080
+    ])
+
+    #: Supported intensity measure component is the geometric mean horizontal
+    #: component
+    DEFINED_FOR_INTENSITY_MEASURE_COMPONENT = const.IMC.AVERAGE_HORIZONTAL
+
+    #: Supported standard deviation type is only total, see table 7, page 35
+    DEFINED_FOR_STANDARD_DEVIATION_TYPES = set([
+        const.StdDev.TOTAL,
+        const.StdDev.INTER_EVENT,
+        const.StdDev.INTRA_EVENT
+    ])
+
+    #: Requires vs30
+    REQUIRES_SITES_PARAMETERS = set(('vs30', 'z1pt0'))
+
+    #: Required rupture parameters are magnitude and top of rupture depth
+    REQUIRES_RUPTURE_PARAMETERS = set(('mag', 'rake'))
+
+    #: Required distance measure is closest distance to rupture
+    REQUIRES_DISTANCES = set(('rrup', ))
+
+    def get_mean_and_stddevs(self, sites, rup, dists, imt, stddev_types):
+        """
+        See :meth:`superclass method
+        <.base.GroundShakingIntensityModel.get_mean_and_stddevs>`
+        for spec of input and result values.
+        """
+        C = self.COEFFS[imt]
+        mean = (np.log(self.get_magnitude_term(C, rup) +
+                       self.get_distance_term(C, dists.rrup)) +
+                self.get_site_amplification(C, sites))
+        stddevs = self.get_stddevs(C, sites.vs30.shape, rup.mag, stddev_types)
+        return mean, stddevs
+
+    def get_magnitude_term(self, C, rup):
+        """
+        Returns the magnitude scaling term in equation 3
+        """
+        b0, stress_drop = self._get_sof_terms(C, rup.rake)
+        if rup.mag <= C["m1"]:
+            return b0
+        else:
+            # Calculate moment (equation 5)
+            m_0 = 10.0 ** (1.5 * rup.mag + 16.05)
```

Comment:
In order to have numbers completely consistent, we should move conversion equations like this one in a utils module.

Reply:
This makes sense to me. There are about 3 - 4 different utils.py files in different places around the engine (one in risklib, one in hmtk/strain, one in hmtk/seismicity) so looking at consolidation (at least of the more generic formulae such as these) is not a bad idea.